org/mozilla/jss/util/JDKCompat.java

```diff
@@ -0,0 +1,21 @@
+package org.mozilla.jss.util;
+
+import java.lang.reflect.Method;
+
+import javax.net.ssl.SSLParameters;
+
+public class JDKCompat {
+    public static class SSLParametersHelper {
+        public static String[] getApplicationProtocols(SSLParameters inst) {
+            try {
+                Method getter = inst.getClass().getMethod("getApplicationProtocols");
+                Object result = getter.invoke(inst);
+                if (result instanceof String[]) {
+                    return (String[]) result;
+                }
+            } catch (Throwable t) {}
```

Comment:
If we want to make ALPN support optional, we should only catch `NoSuchMethodException` and explicitly ignore or a generate warning. All other exceptions should still be propagated to the callers.
```
} catch (NoSuchMethodException e) {
    // ALPN not supported
} catch (Exception e) {
    new RuntimeException(e);
}
```
Also, it's not necessary to check the type of `result` since it's always `String[]` in all Java versions (if the method is available).

Reply:
I think we need to catch more than `NoSuchMethodException` -- we also need to handle what happens when reflection fails (e.g., `SecurityManager` fails to allow the operation). In this case, we should log a message saying why we can't proceed and tell them to use `JSSParameters` if it is really serious.

Reply:
I think since we're determining ALPN support based on the availability of the method, we should only check `NoSuchMethodException` and ignore that exception only. If throws a different exception, that would be a different issue that needs to be reported to the caller. If we ignore all exceptions, that means ALPN support is really arbitrary (i.e. you can use it if it works) and there's no way to guarantee it if you need it.

Reply:
I'm worried about what happens [when a `SecurityException` is thrown](https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#getMethod-java.lang.String-java.lang.Class...-).

I don't think we can guarantee that the method _does_ exist when a `SecurityException` is thrown. So we don't have enough information as to whether or not we should treat this as fatal or not. I'd be fine limiting it to `NoSuchMethodException` and `SecurityException`.

Reply:
We only have this problem only because we need to store a value in `appProtocols`, and a `null` could be interpreted as real `null` or unset. See my other comment about catching the `NoSuchMethodException` in `JSSEngine` so we will not have this issue.